Typed Command with multiple input values to replace SignalX

burkeds opened the pull request:
Addresses issue #1087 

New Command with R, RW, W, and X flavours.

SoftCommandBackend enforces typematching between callback signature and command_args. command_args must be a list of types in order of the callback signature.

MockCommandBackend included for testing.

Example use of soft signal.

```python
from ophyd_async.core import CommandRW, soft_command_rw

my_async_command: CommandRW[[int], float]

async def async_compute(x: int) -> float:
    return float(x) * 2.0

my_async_command = soft_command_rw(
    command_args=[int],
    command_return=float,
    command_cb=async_compute,
    name="async_compute_gain",
)

await my_async_command.call(1)
```

burkeds committed:
First draft of Command classes. WIP

burkeds committed:
Merge branch 'main' into command

burkeds committed:
Developing unit tests

burkeds committed:
fixing datakey

burkeds committed:
Core command

burkeds committed:
merge

burkeds committed:
Core commands with syntax fixes. Removed old tango command code

burkeds committed:
Added test for behaviour when command is called with kwarg ordering different than signature and command_args list.

burkeds committed:
linting

burkeds committed:
removed ... and using pass

burkeds committed:
Merge branch 'main' into command

coretl commented:
Actually if you don't care about keyword args you could do:
```python
class MyDevice(TangoDevice):
    my_command: Command[Callable[[int, str], Awaitable[float]]]

a_float = await MyDevice().my_command.trigger(0, "foo")  # type checks as float
```
but I can't see how to avoid having to put `Awaitable` in the type string.

burkeds committed:
Merge branch 'main' into command

burkeds committed:
Merge branch 'main' into command

burkeds committed:
Merge branch 'command' of https://github.com/bluesky/ophyd-async into command

burkeds committed:
wip

burkeds committed:
Improved command with better static typing

burkeds committed:
Telling docs to ignore ParamSpec and TypeVar in _command

burkeds committed:
Merge branch 'main' into command

burkeds committed:
removed custom error classes. Added device logging

burkeds committed:
shortened source signature

burkeds commented on src/ophyd_async/core/_command.py:
```diff
@@ -0,0 +1,340 @@
+                raise ExecutionError(f"Command execution failed: {exc}") from exc
+
+    def _async_lock(self):
+        return self._lock
```
The lock is used in call. If multiple soft signals use the same callback there is the possibility of race conditions. The lock is intended to prevent that.

burkeds committed:
Removed unnecessary factory functions. Removed unused units and precision arguments

burkeds commented on src/ophyd_async/core/_command.py:
```diff
@@ -0,0 +1,340 @@
+
+    @abstractmethod
+    def source(self, name: str, read: bool) -> str:
+        """Return source of command."""
```
done

burkeds committed:
Refactored SoftCommandBackend to enforce annotations at runtime. This means we can rely on annotations of the callback

burkeds committed:
Merging main

burkeds committed:
Refactored call to triggerable

burkeds committed:
Added runtime validation for Array1D and Sequence types

coretl commented:
Unfortunately to support keyword specification of parameters I think we would need to make it generic on a callable, rather than on `P` and `T`. Something like:

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBAxmA28CmMAuBLMA7AzgOgEMAjGKdCABzBFSgEEB3A9VY5AGigGEDE2kAUKEhRUATwrosAczKVqtAOJIsSEOhicACuFRg48TgBUJSAGoEQAgVygBeKCYrnLACgBEXd5yJgArlgAJnY8fETIAJTWAjDwBDg43JAQBEGuyqrqMADaXAC6EQBcAlClUIFIwFAA%2BtVSLLWuOEjwwJzAATCF3EVQeP3WZTXVMLzwtd1c0fFiWKQVVcCuAB7dUqicYt04qCARUAC0AHxQwPBgBKjd-XjRAMRQAMrOMOjAYlBEH1Q4LJhYMQAjJNkqlArkxvxstl1pwdiA8pxGMxWOEkNkzhdUHkcfYkhAUmlgFFrA9nig3h8vlAANZMEDSHAxOIJKAAMVcOjAegMRRKZRmc3KlWGo0QjWarU4qzIWA2UC2UHh%2B2Op3Ol2uA2sMAATCCCWDsmy8niuKCiSSBIL5iLUEgdq4%2BUNloC8QQmCxYIDXABmTjuYBgMDuKLOnVuj20XUrOx%2BhV2ANBkMCIA)

```python
from collections.abc import Awaitable, Callable
from typing import Generic, Protocol, TypeVar

C = TypeVar("C", bound=Callable)


class Command(Generic[C]):
    def __init__(self, func: C): ...

    __call__: C


async def f(x: int, y: str) -> float: ...


# Specify by position
c1: Command[Callable[[int, str], Awaitable[float]]] = Command(f)


# Specify by kwargs
class F(Protocol):
    async def __call__(self, x: int, y: str) -> float: ...


c2: Command[F] = Command(f)


async def test():
    x1 = await c1(3, "foo")
    x2 = await c2(x=3, y="foo")

```

burkeds committed:
Merge branch 'main' into command

burkeds committed:
Made Command a Generic. Using call again instead of Trigger but __call__ still returns a status object. wait_with_error_message in AsyncStatusBase now returns the value of the awaitable so that is it accessible from status.task.result()

burkeds committed:
Added _call to avoid AsyncStatus.wrap. Using AsyncStatus.wrap changes the return typing of __call__ from T to AsyncStatus.

burkeds committed:
Testing out tango commands

burkeds committed:
Placeholder command connector

burkeds committed:
Made tango_command accept a DeviceProxy so a new DeviceProxy does not need to be created

burkeds committed:
wip

burkeds commented on src/ophyd_async/core/_command.py:
```diff
@@ -0,0 +1,349 @@
+        """Connect to underlying hardware."""
+
+    @abstractmethod
+    async def __call__(self, *args: P.args, **kwargs: P.kwargs) -> T_co:
```
Covariance is useful I think in the case of Enums. This allows the backend to use subtypes of StrictEnum. Here is an example:

https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDiApikSEgMYA0UAKokQGoCGIAUKJFCQK7RY48UAKIo%2BNZjxhg2bAMRQAwmABurJMxT54CIlDXlmAIwA2RNrQD6FMFAC8dBixAAKAERWbbmjYMatdrQgPEQAlPJ01rYO9LrO7rRu4REAIkTAqHpgPCDcYtAAFkhkrBQFcFA8AM6oGDAFegjgKkgAJkStUEbMVXoUJj29VWz9g1AAQj1EonwuMxChAFxsUKtQCIOyo1VVUADKMCDzLlWHNJO980sra253B%2BToeXx3brIKiiBEzDB9YCgUL6-Z4QYbbXbKEy4FwPK7LNZQABKwhS9igkmkLnCCIIyOEADk0RiwFitgMdhNyGgCjBIdDYflrgj8cIAPKEhzE0kIgBiABkAKqsolSEnJBTEUjkChQGxaZiZdjgiZgAAeLklZEoAG1PGAALpMtbtYBQSyWVBIGDmk5EEzAGhqEwhRaRGxGhGrXr2gB0TpCaP95giPJ4AJgSH%2B3FVugoEae4zV2oZfH1bBN60e1t4EBcRjVrsTquTh3mhqgAFoAHxQfH-IjwtZNVAwFzANzzfTMZ0NqAAb3zqr93ZCAF8ku8lF8fnpB0RdgB3K0FKCtJDAYBkEj4HOwBhgsBQkCWQeFpN0kD6tFFlwXn3IlLhIxUmkngsq4vjF%2B0w%2B4K8OG8vyQakfyPH0WXZcU6CKRdcAAa12IwiAoSRemWBQAEYfQ-TBdl8dRNHwVA3RkBQACZsIvdEUE6ICQKo1hZzAep9lLfIIgAZmwlN%2BF2eo-ggCAo26Xo910Nhmy0SwcxcGwjzfVVQk9KAFFZKQECkV1OyDV07wfCSs2k-I82-BSlIRVT1M0kR8i7HtC2-O8IPxIA

coretl replied:
Thanks for explaining, that makes sense

coretl commented on src/ophyd_async/core/_command.py:
```diff
@@ -0,0 +1,349 @@
+        self.log.debug(f"Command {self.name} returned {result}")
+        return result
+
+    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> AsyncStatus:
```
Yes, the type hinting I proposed in https://github.com/bluesky/ophyd-async/pull/1138#issuecomment-3841598833 will only work if `CommandBackend.__call__` has the same signature as `Command.__call__`.

If we do this then we need to add a return value to the `Status` protocol. 
@tacaswell what do you think?

burkeds committed:
__call__ now uses the same converters as signal. inspect.signature.bind is used to validate callback arguments against its signature before being passed to the callback

burkeds committed:
Moved converter creation to init

burkeds committed:
Merge branch 'main' into command

coretl commented:
> @coretl per our conversation in the dev call, should I add a subclass of AsyncStatus which returns its value? If so, what should I call it?

I think we should prototype this with plain async functions all the way to a Tango implementation, then revisit at the end to see if we want to make it `AsyncStatus`. I'll raise an issue on bluesky in the meantime to discuss and link it here.

burkeds committed:
AsyncStatus no longer returns a value

burkeds committed:
Merge branch 'main' into command

burkeds committed:
In AsyncStatus, the value returned by the awaitable can be accessed using the value property

burkeds commented:
> @burkeds I've written [bluesky/bluesky#1988](https://github.com/bluesky/bluesky/issues/1988). I think we should add `value` to the base `AsyncStatus` and make it always return `self.task.result()` which will do for now.

Done. I made the Awaitable in AsyncStatusBase generic over T since the value property now is typed to return something.

coretl commented:
Please could you change to `execute` from `__call__`? Then I'll be interested to see how the tango implementation goes. I've updated the playground link:

Code sample in [pyright playground](https://pyright-play.net/?code=JYWwDg9gTgLgBMAdgZzAUwMYwFADMoQhwYQA2pmMwEKAdAIYBGGC408AggO73AxMUANHADC9cgLR4CRGAE8wSAOatIsOB0RzhImhnow0iA2mEBxI2ijAMwgAr0o9EAGV0tuHYIwIJUsIAVBTQANUdsbDs4AF5PR2c3TAAKACI7FIBKbACYuCD0MKhUgMyI7AxSemRkUUIQekQAEwAhegwAayNGpItEKxsAbTtAgF0MgC5sOGm4RrRcOGQIAFcoDDQk5DRSXGFjEDRxxZgoDLgAWgA%2BY6gj2nupmaq5RBY5hbQAD0xlw03t3ZwABUjiUyCOdgYUDBwiBQPaPGh4M8tARoOQZyueTuDwiFSqNRcEFwMF0IHqTVaHS6STJFJabU6TSGowmj2m7zgAH0uUg%2BDz-jthLhlq8jmIJIwKCyNDw%2BJIBgERmNJjM1YsAbQRa9ctqMBE1c8dZyvj8-lshcD0RCoTDgfDEWCbWikZjrgFVerplA0DBVog4PQ5fALbgtaKMEkQUjYQ70Vk8ZVqnkGkoIHSGgzqU1aXVM1SmY0ZUq2WrOTy%2BTABaHhGBvL4yFz9OQuXrxeJKlK0ANdK8TMZDANNNpPCNhNxePwu4rlaWvRqdrRkMAlAPVmhckhUJQlyu1z6knWID4-E2O62I1l53AAMRwAByaDQjTgPjgkDAC4WK3gxNfAAtgGQA0nmQF43nmOBTQwX4NhrK0kRtdFYxdJ0UVQjELndHFaDKfFkwzJovGPBtSCSYiTzIOcOUgpZVnWQVcDdG4cLwpMakI7pen6DBixVdlZkgitEH5LlGOERhGS6cU80pKTmWGPJ%2BOvUNaEk7MX1idTCxAmiFjotY4IBZjkBOT0vR9P0oADVTtK6JcVkM1J9jQUoBKNCCPm%2BGDzQBWFrRRZD7Qw51HUwrEPQEtU6yQGBUgAEQgRZCA3UgICUJRlFKa9LP9QNgy-NT5MaWhoNgqMgrhDCE2wDzBIWXAkk%2BI5YuEOQjlM04sLgXA0oMcy4Fy6yer6uLPjgABqEaIAMJI5AyGrOSWEkm1k7o21EDsFUUid5WnJUxm6zi%2BIGoaA04pIiRJTiCxpPUFrKTl%2BEQNNVvJTMkgEo9KNIM8Ww2iVO2lXt9EMAdu2Hewx1lScFQOwRsGY47FKVU7fTyi6AlTdM1tunNvtIv7fvumq7xAOQ5gANxsNBaDAOQyjvRIMGAXA5DgRh2Yw8p2LgAAxXN3qI%2Bs-HsEWqIGuqTW88r4OahBEBgNqOpOZjepmmBWMTAk4AAWTkeK0Gp9YBowABGGShaLAZWpuaH1YMEZpiZ9xWfZzn3wgZcqBoASMAAJiOPnnbgZm3Y5rmwsZqCmjgcmqZpumGYiKXIMMUyuXjo2aaSajb1Ef9MHaAC0HZ%2BoYAwf9X2CYCyy5Zb4FifXDeNjYrxmRp6%2BJGBaHN3IG7e%2BkkiYgTO4b3v-f77vB4%2Bke6%2BetNcmb7OGPbjkuQXiBe7N3JN5nnM%2BdKmXDDX2YN%2Bxifd%2Bx-fukPsqT4E3BoFmBW4AGMfu%2BETvN5GAbDWqKw8AgyTlmNvI%2BZoNhm2ECkWgZtMgxFiGbWBUVQJbHUMAvgoCA7gJ8hsAAzNA2g-t4HRFiHgohKDpgEkAflEBJVsH3w2J8aIUC4ByGiDA4hZxSFwCQf7ShgYAHqHoUggyq8BHUOERPBy9E27YCAA)

```python
import inspect
from collections.abc import Awaitable, Callable
from typing import Any, Concatenate, Generic, ParamSpec, Protocol, TypeVar

P = ParamSpec("P")
T = TypeVar("T")


class CommandBackend(Generic[P, T]):
    def source(self, name: str) -> str: ...
    async def execute(self, *args: P.args, **kwargs: P.kwargs) -> T: ...


class SoftCommandBackend(CommandBackend[P, T]):
    def __init__(self, func: Callable[P, Awaitable[T]]):
        self.func = func

    async def execute(self, *args: P.args, **kwargs: P.kwargs) -> T:
        return await self.func(*args, **kwargs)


class TangoCommandBackend(CommandBackend[P, T]):
    def __init__(self, protocol_call_func: Callable[Concatenate[Any, P], Awaitable[T]]):
        self.signature = inspect.signature(protocol_call_func)
        # Need to pop self out of this

    async def execute(self, *args: P.args, **kwargs: P.kwargs) -> T: ...


class CommandProtocol(Protocol):
    def source(self) -> str: ...


class Command(Generic[P, T]):
    def __init__(self, backend: CommandBackend[P, T]):
        self.backend = backend

    def source(self) -> str:
        return self.backend.source("name")

    async def execute(self, *args: P.args, **kwargs: P.kwargs) -> T:
        print("Do some logging")
        return await self.backend.execute(*args, **kwargs)


async def f(x: int, y: str) -> float:
    return float(x + float(y))


def soft_command(func: Callable[P, Awaitable[T]]) -> Command[P, T]:
    return Command(SoftCommandBackend(func))


def tango_command(
    protocol_call_func: Callable[Concatenate[Any, P], Awaitable[T]],
) -> Command[P, T]:
    return Command(TangoCommandBackend(protocol_call_func))


# mydevice.py


# Specify by kwargs
class F(CommandProtocol, Protocol):
    async def execute(self, x: int, y: str) -> float: ...


class MyDevice:
    c1: Command[[int, str], float]  # Specify by position
    c2: F  # Specify by kwargs


# end mydevice.py


async def test_mydevice():
    # Check they match types
    d_soft = MyDevice()
    d_soft.c1 = soft_command(f)
    d_soft.c2 = soft_command(f)
    d_tango = MyDevice()
    d_tango.c1 = tango_command(F.execute)
    d_tango.c2 = tango_command(F.execute)
    for d in [d_soft, d_tango]:
        assert await d.c1.execute(1, ".1") == 1.1
        assert await d.c2.execute(3, ".2") == 3.2
        assert await d.c2.execute(x=1, y=".2") == 1.2
        assert d.c1.source()
        assert d.c2.source()

```

coretl commented on src/ophyd_async/core/_command.py:
```diff
@@ -0,0 +1,259 @@
+
+
+def soft_command(
+    command_cb: Callable[P, T | Awaitable[T]],
```
I think we want a callable or async callable rather than a callable that could return either...

```suggestion
    command_cb: Callable[P, T] | Callable[P, Awaitable[T]],
```

burkeds committed:
moved _wait_for to _utils for use by command and signal

burkeds commented on src/ophyd_async/core/_command.py:
```diff
@@ -0,0 +1,259 @@
+    try:
+        return await asyncio.wait_for(coro, timeout)
+    except TimeoutError as exc:
+        raise TimeoutError(source) from exc
```
done

burkeds committed:
Moved T_co to _utils. Using P from _utils in Command

burkeds commented on src/ophyd_async/core/_command.py:
```diff
@@ -0,0 +1,259 @@
+            result = self._command_cb(*bound.args, **kwargs)
+            if inspect.isawaitable(result):
+                result = await result
+            self._last_return_value = cast(T, result)
```
I think it could be useful. There could be some scenarios where you want to know the last value returned by a command without running it. 

It is probably not very useful in a typical plan because you can simply store the last value if you want it. But it is more useful in the context of "ophyd as a service" because now the last value is accessible in the device.

I would be of a mind to keep it for now.

burkeds committed:
Refactored MockCommandBackend to use execute_mock and more closely resemble MockSignalBackend

burkeds committed:
Update src/ophyd_async/core/_command.py

Co-authored-by: Tom C (DLS) <101418278+coretl@users.noreply.github.com>

burkeds committed:
Merge branch 'command' of https://github.com/bluesky/ophyd-async into command

burkeds committed:
Telling docs to ignore _utils.T_co

burkeds committed:
Refactored Command to use execute instead of __call__

burkeds committed:
Removed last_return_value from SoftCommandBackend

coretl commented on src/ophyd_async/core/_command.py:
```diff
@@ -0,0 +1,234 @@
+        """Returns the source of the command."""
+        return self._connector.backend.source(self.name)
+
+    async def _execute(self, *args: P.args, **kwargs: P.kwargs) -> T:
```
Any reason this couldn't be:

```suggestion
    @AsyncStatus.wrap
    async def execute(self, *args: P.args, **kwargs: P.kwargs) -> T:
```

Probably the type signature of `AsyncStatus.wrap` needs changing a bit to support it

coretl commented on src/ophyd_async/core/_command.py:
```diff
@@ -0,0 +1,234 @@
+        execute_mock = AsyncMock(
+            name="execute",
+            spec=Callable[P, Awaitable[T]],
+            side_effect=self._mock_execute_callback,
```
Probably want to add something like this here:
https://github.com/bluesky/ophyd-async/blob/beba1c88da43adee62db2675facb5682a024f25a/src/ophyd_async/core/_mock_signal_backend.py#L68-L70

but we need to manufacture a value to return. If we made a converter for the return value type like the input types then we could use `converter.write_value(None)` to make it for us like:
https://github.com/bluesky/ophyd-async/blob/beba1c88da43adee62db2675facb5682a024f25a/src/ophyd_async/core/_soft_signal_backend.py#L141

coretl reviewed (commented):
Looking good, couple of minor points. I'm interested to see how the TangoCommandBackend looks, then I'll have a go at the EPICS one...

burkeds committed:
Simplified execute with AsyncStatus.wrap

burkeds committed:
merging

burkeds committed:
Added a converter for MockCommandBackend so it can return a value. In order to do this, MockCommandBackend needs to know the type to return. This necessitates a method in CommandBackend which can be used to retrieve the type Mock should use. A new abstract method get_return_type was added to CommandBackend to accomodate this need. SoftCommandBackend now stored the return type from the callback signature which can then be retrieved by this method.

burkeds committed:
Merge branch 'main' into command

burkeds committed:
Improving Mock test coverage

burkeds committed:
Merge branch 'main' into command

burkeds committed:
Merge branch 'main' into command

burkeds committed:
WIP. need to refactor device filler

burkeds committed:
testing device filler

burkeds committed:
Probably getting an error here because state is both an attribute and a command

burkeds committed:
First working draft of TangoCommandBackend. Required a refactor of _tango_transport especially get_python_type to properly handle return types. CommandProxy refactored to no longer use the command character enum. Use of this enum should be removed completely in the future. WIP, need to fix older tests for new command structure

burkeds committed:
Adding trigger for backwards compatibility with SignalX

burkeds committed:
Added trigger method to Command for backwards compatibility with SignalX. Trigger calls execute with no args and does not return a value. Refactored tango tests to not use SignalX. Removed some cmd tests. Refactored tests for get_python_type to use better expected types.

burkeds committed:
passes linting. Had to make copies of some utilities in tango.core._utils to avoid accessing private core._utils.

burkeds committed:
tigger calls execute with explicitly no arguments

burkeds committed:
Merge branch 'main' into command

# Conflicts:
#	src/ophyd_async/core/_device_filler.py

burkeds committed:
fix linting

burkeds committed:
fixing device_filler

burkeds committed:
Fixing type-checking issues

burkeds committed:
linting

burkeds committed:
fixing docs

burkeds committed:
Fixed issue with tango command types not being converted correctly

burkeds committed:
Merge branch 'main' into command

burkeds committed:
Improving coverage

burkeds committed:
removed commented out code

burkeds committed:
Merge branch 'main' into command

coretl reviewed (changes requested):
A couple of structural things, I haven't tried to understand the tango specific parts yet, but I'll have a look at those bits shortly

burkeds committed:
Remove accidentally committed temp directory

burkeds committed:
removed get_return_value in favour of simply using the datatype attribute similar to SignalBackend.

burkeds committed:
removing get_return_type from tango backend

burkeds committed:
test now import make_converter from _soft_signal_backend so it does not need to be exported.

burkeds committed:
Leaving choice to raise NotConnectedError to the backend

burkeds committed:
trigger now calls the backend execute directly

burkeds committed:
Refactored Command.trigger() into a new TriggerableCommand class. This avoids having one class which covers the general case of typed args/return and the special case of no args. Trigger will fail if the backend expects arguments. It will not return any values from the backend.

burkeds committed:
Tango connector and device filler now infer None/None commands as TriggerableCommand instead of simply Command

burkeds committed:
Removed commented code

burkeds committed:
fixing tests

burkeds committed:
Added test for tango_command factory. factory can now return TriggerableCommand.

burkeds committed:
Added TriggerableCommand to tango_command return typehint

burkeds committed:
TangoCommandBackend now takes an inspect.Signature object instead of type[T]

burkeds committed:
linting

burkeds committed:
Fixing linting and type checking. Pyright said to replace T with object in tango_command because it only appears once in the signature

burkeds committed:
In procedural constructions, tango commands will work if the datatype is not specified. This fails for declarative constructions

burkeds committed:
fix type checking

burkeds committed:
linting

burkeds committed:
Merge branch 'main' into command

burkeds committed:
Added parameter type checking for TangoCommandBackend. Parameter types are stored in the backend as param_types, a list of SignalDatatype. Passing None as the parameter type bypasses the check in the procedural approach but fails in the declarative approach to signal declaration.

burkeds committed:
removed redundant use of _wait_for in Tango backend

burkeds committed:
removed _wait_for from tango._utils

burkeds committed:
Added some null protection to fill_child_signal and fill_child_command

burkeds committed:
Passing a signature to tango_command does not allow for type hinting. To obey type hinting we need to define a callable stub function and pass that as call_spec to tango_command

burkeds committed:
added tango_triggerable_command which returns TriggerableCommand. tango_command can no longer return a TriggerableCommand

burkeds committed:
datatype is not an appropriate attribute for Command as it is for signal because the whole function signature is relevant to the backend. This refactors Command and CommandBackend to instead store a signature attribute which is expected to be an inspect.Signature object. The TangoCommandBackend then validates according to this signature on connect.

burkeds committed:
linting and type checking

burkeds committed:
Merge branch 'main' into command

burkeds committed:
Fixing tests

coretl commented on src/ophyd_async/core/_command.py:
```diff
@@ -0,0 +1,271 @@
+    get_origin,
+    get_type_hints,
+)
+from unittest.mock import AsyncMock
```
We already pull this in via an init import for signals, so this is no different

coretl commented on src/ophyd_async/core/_command.py:
```diff
@@ -0,0 +1,271 @@
+)
+
+
+class CommandBackend(Generic[P, T_co]):
```
Not required as static analysis does this for us

coretl commented on src/ophyd_async/core/_command.py:
```diff
@@ -0,0 +1,271 @@
+            spec=Callable[P, Awaitable[T]],
+            side_effect=self._mock_execute_callback
+            if self._mock_execute_callback
+            else lambda *args, **kwargs: None,
```
Now we have a signature I think we can do better and check the args and kwargs, then compute a default return value using the converter for the return type

burkeds committed:
Many minor improvements in response to Claude review

burkeds committed:
Merge branch 'main' into command

burkeds committed:
Merge branch 'main' into command

coretl committed:
Added EPICS support for TriggerableCommand

Also added ADR, and some typing tweaks

coretl committed:
Merge remote-tracking branch 'origin/command' into command

coretl committed:
Fix PVI to use TriggerableCommand instead of SignalX

coretl committed:
Fix remaining tests

coretl commented:
@burkeds I think this is ready, please can you review?

burkeds reviewed (commented):
@coretl I don't see any major changes to the tango side of things. Looks good to me!

coretl committed:
Reproduce problem

coretl commented:
I'll merge when 0.17 and 0.18 have been released:
https://github.com/bluesky/ophyd-async/issues/1016

burkeds committed:
Merge branch 'imports' into command

burkeds committed:
Removed TestConfig from tango.testing. Defined two new protocols in tango._utils called AttributeConfig and CommandConfig to check objects against. This should prevent the need to import test machinery when using the tango backend.

burkeds committed:
Merge branch 'main' into command

burkeds commented:
I merged `imports` into `command`. I removed `tango.testing.TestConfig` and replaced it with two protocols (`AttributeConfig` and `CommandConfig`) in `tango._utils`. Config-like objects are now tested against these protocols in the tango backend.

coretl committed:
Merge main into command, resolve conflicts with TriggerableCommand

coretl committed:
Remove dead timeout param from CommandProxy.put; rename ADR to 0018

coretl committed:
Docs: document Command and TriggerableCommand, update tutorial and docstrings

coretl committed:
Replace SignalX with TriggerableCommand in Eiger detector

coretl committed:
Merge remote-tracking branch 'origin/main' into command

coretl committed:
Fix docs
